asynciterator.ts

```diff
@@ -1923,3 +1923,116 @@ type SourceExpression<T> =
 
 type InternalSource<T> =
   AsyncIterator<T> & { _destination: AsyncIterator<any> };
+
+interface Transform { filter: boolean, function: Function }
+
+function build(transforms: Transform[]) {
+  return transforms.reduceRight((f, transform) => transform.filter ?
+    (item: any) => transform.function(item) ? f(item) : null :
+    (item: any) => f(transform.function(item)), (e: any) => e);
+}
+
+/**
+  An iterator that is used to quickly transform items, it is optimized for performance
+  and uses *mutations* - this means that it is unsafe to do anything with the original iterator
+  after map, transform and syncTransform have been applied to it.
+  @param source The source to transform
+  @extends module:asynciterator.AsyncIterator
+*/
+export class FastTransformIterator<T> extends AsyncIterator<T> {
+    private transforms: Transform[] = [];
+    constructor(private source: AsyncIterator<T>) {
+      super();
+      source.on('readable', () => {
+        this.emit('readable');
+      });
+      source.on('end', () => {
+        this.close();
+      });
+    }
+
+    read(): T | null {
+      const func = build(this.transforms);
+
+      this.read = () => {
+        const { source } = this;
+        let item;
+        while ((item = source.read()) !== null) {
+          if ((item = func(item)) !== null)
+            return item;
+        }
+        return null;
+      };
+
+      return this.read();
+    }
+
+    /**
+      Filter items according to a given function
+      @param {Function} filter The function to filter items with
+    */
+    filter<K extends T>(filter: (item: T) => item is K): FastTransformIterator<K>;
+    filter(filter: (item: T) => boolean): FastTransformIterator<T>;
+    filter(filter: (item: T) => boolean): FastTransformIterator<T> {
+      this.transforms.push({ filter: true, function: filter });
+      return this as unknown as FastTransformIterator<T>;
+    }
+
+    /**
+      Maps items according to a given function
+      @param {Function} map The function to map items with
+    */
+    map<D>(map: (item: T) => D): FastTransformIterator<D> {
+      this.transforms.push({ filter: false, function: map });
+      return this as unknown as FastTransformIterator<D>;
+    }
+
+    /**
+      Transforms items according to a synchronous generator (hence no need for buffering)
+      @param {Function} transform The function to transform items with
+    */
+    syncTransform<D>(transform: (item: T) => Generator<D>): FastTransformIterator<D> {
```

Comment:
What are the use cases for this transformation? Is this just a multimap?

Reply:
Yep!